soc: xtensa: partial fix of socket misuse and refine the code

enjiamai opened the pull request:
This PR fix bug and refine the client-server CAVS tool, includes:

- Improve the firmware transfer reliability by fixing the misuse of the socket.  Fix the most frequent occurence of the common `recv()` bug described here: https://docs.python.org/3/howto/sockets.html#using-a-socket
The longer term fix is to switch to a higher level API like Python Remote Objects.

- Not rely on the client's command to disconnect. Previously we rely on the SIGINT to send stop_command to the server, but it does not work well in some environments. Refine the whole logic and the sever disconnect service by checking if the client is alive or not. And also remove the stop_command code.

Fixes #46864

PS:  I split the port supporting to anther PR: https://github.com/zephyrproject-rtos/zephyr/pull/46972

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -115,7 +115,20 @@ def main():
+        PORT_LOG = int(h[2])
+        PORT_REQ = int(h[1])
+
+    elif l == 2 and l > 3:
```
Guess you wanted `or` here, no?

enjiamai replied:
Thanks for the correction, it shall be OR here, fixed.

edersondisouza commented on scripts/west_commands/runners/intel_adsp.py:
```diff
@@ -124,4 +125,8 @@ def flash(self, **kwargs):
         # make the log output to the PTY.
         if not self.pty:
-            self.check_call(self.log_cmd)
+            # Though it's encouraged to do this way due to
```
I'm confused, commit message says something is _not_ encouraged, here it is. I guess both messages could be clearer.

enjiamai committed:
soc: xtensa: partial fix of socket misuse and refine the code

1. Improve the firmware transfer reliability by fixing the misuse
of the socket. Fix the most frequent occurence of the common `recv()`
bug described here:

https://docs.python.org/3/howto/sockets.html#using-a-socket

The longer term fix is to switch to a higher level API like Python
Remote Objects.

2. Not rely on the client's command to disconnect. Previously we
rely on the SIGINT to send stop_command to the server, but it does
not work well in some environments. Refine the whole logic and the
sever disconnect service by checking if the client is alive or not.

These changes make the client-server-based cavstool more stable.

Fixes #46864

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

laurenmurphyx64 reviewed (dismissed):
Tried this out with `intel_adsp_cavs25` with `tests/kernel/common`, and it's awesome!!! :rocket: :heart: Since the fix works so well, I don't see any need to follow my suggestion about SCP in https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1165742158. It's great that you could keep it all in Python with TCP :+1:

enjiamai commented:
Thanks for verifying it! @laurenmurphyx64 , I also tried running multiple boards at the same time in the network which using the port mapping, so far it works well. :)

andyross reviewed (dismissed):
No complaints about the code here.  Just a general comment:

I do worry that as cavstool grows features (first to add the test integration support, now the new client/server architecture) that we're obscuring the core code that does the loading behind a big facade of other stuff that also has to be run In Just The Right Way to make things work.  That impedes development flow.  It might be time to split the code up so that a minimal "just run this file please" tool can continue to be maintained.

Let's all remember that cavstool isn't the product.  Firmware in production will never use this and all final testing has to happen in a SOF binary with the kernel loaders.

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +661,69 @@ async def _main(server):
+        md5_rx = hashlib.md5(total).hexdigest()
+        md5_tx = md5_tx_b.decode('utf-8')
+
+        if not md5_tx == md5_rx:
```
Why not `md5_tx != md5_rx`? Maybe it's just me, but current way is kinda hard to read.

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +661,69 @@ async def _main(server):
+        md5_tx = md5_tx_b.decode('utf-8')
+
+        if not md5_tx == md5_rx:
+            log.error(f'md5 dismatch: {md5_tx} vs. {md5_rx}')
```
s/dismatch/mismatch/

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -694,24 +735,21 @@ def handle(self):
+                self.request.sendall("failed".encode('utf-8'))
+                log.error("Receive firmware failed.")
+
+            log.info("Flash completed. Ready to run...")
```
I'm kinda confused, only the download was completed at this point, no? No flashing, like, writing firmware to memory where CPU will start running it, right?

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -50,18 +54,51 @@ def send_cmd(self, cmd):
+
+        # Pack the header and the expecting packed size is 80 bytes.
+        # The header by convention includes:
+        # size(2), filename(42), MD5(32) and pad(2)
```
I'm kinda confused by this math: 2 + 42 + 32 + 2 == 78, not 80. What am I missing?

enjiamai replied:
Sorry for my carelessness.  Correct them in both client/sever comments. Thanks!

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -50,18 +54,51 @@ def send_cmd(self, cmd):
 
     def download(self, filename):
```
I know the ship has sailed, but isn't this... "uploading" the file?

enjiamai replied:
En... you are right, The client side uploads and the server side downloads the FW.  Fixed it.

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -50,18 +54,51 @@ def send_cmd(self, cmd):
+        values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'), flen)
+        log.info(f'filename:{fname}, size:{fsize}, md5:{md5_tx}, len {flen}')
+
+        s = struct.Struct('I 42s 32s H')
```
Oh, it seems size is actually 4, from the format string.

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -50,18 +54,51 @@ def send_cmd(self, cmd):
-            ret = self.sock.sendfile(f)
+            log.info(f'Sending...')
+
+            m = packed_data + f.read(BUF_SIZE - packed_size)
```
Nitpick: i'm not sure what `m` is here, maybe a more meaningful variable name?

enjiamai replied:
Replaced "m" it with "first_send". It's more meaningful. thanks!

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -89,8 +126,8 @@ def main():
-        finally:
-            cleanup()
+        #finally:
+        #    cleanup()
```
Why this commented out code?

enjiamai replied:
Originally I want to disable in the second commit then remove it in the third commit. Thanks for pointing this out. Got it fixed.

edersondisouza commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -50,18 +54,51 @@ def send_cmd(self, cmd):
-        log.info(f"Sent fw:     {send_fn}, {filename}")
 
         self.download(filename)
+        log.info(f"Done Transfer")
```
Why this extra message? And why no message a few lines above, in the case the `success` reply is not received?

enjiamai replied:
Ok, I remove it here and put a log.error in above code where did not receive successfully. You are right, thanks.

edersondisouza commented:
My new comments are just nitpicks, that could be addressed when code is later revisited to separate the client/server model from the loader.

marc-hb reviewed (commented):
Thanks for the tentative fixes, I did experience very frequent serialization problems and crashes with the new remote feature. I spent a few hours debugging them and it looks like the custom protocol was simply missing delimiters or length fields. I will test this PR today.

I have really not reviewed the code but it already seems clear low-level serialization should not be re-invented here. This is a very common need and there are off-the-shelf solutions for this that avoid typical bugs and save a lot of time. After a 5 minutes search I found:

- Python  remote objects https://pyro4.readthedocs.io/en/stable/tutorials.html
- https://docs.jsonrpc.ru/quickstart.html#a-minimal-application

There may be more options.

If this actually fixes serialization crashes then it should merged, not blocked waiting for the Great Refactor and Simplification that drops most of the code but definitely something to keep in mind.

cc: @teburd

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,69 @@ async def _main(server):
+        fsize = unpacked_data[0]
+        fname = unpacked_data[1]
+        md5_tx_b = unpacked_data[2]
+        pad = unpacked_data[3]
```
Thanks for the pointing this out. @marc-hb, very good suggestion. It can be more simplified. Got learn!

enjiamai commented:
> My new comments are just nitpicks, that could be addressed when code is later revisited to separate the client/server model from the loader.

Hi, @edersondisouza .  I appreciate your review and correct the many errors in the code!  :)

enjiamai commented:
Hi @andyross , @edersondisouza , @laurenmurphyx64 ,  @marc-hb  could you please help on reviewing it again when you got a change? I tested it for one full cycle and the code logic did not change at all. thank you!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,65 @@ async def _main(server):
+
+            total += data
+            l = len(data)
+            cnt = cnt + l
```
Unnecessary, C-style house-keeping, see above. Python objects know their size.

enjiamai replied:
Yes, got it fixed. Thanks!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -41,27 +43,57 @@ def send_cmd(self, cmd):
+            total = len(first_send)
+
+            while True:
+                l = f.read(1024)
```
This seems to be re-implementing https://docs.python.org/3/library/os.html#os.sendfile and https://docs.python.org/3/library/socket.html#socket.socket.sendfile

This is Python, not C: every time you're writing basic code like this there is a 99.9% chance you're re-inventing some library function.

enjiamai replied:
Good suggestion! Simplify this part by using socket.send_file, thanks.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,65 @@ async def _main(server):
+            if not data:
+                break
+
+            total += data
```
Just use a very large BUF_SIZE (16MB) and read once. Keep it simple. If you look at the size of the files in https://github.com/thesofproject/sof-bin/tree/main/v2.2.x/sof-v2.2-rc1/intel-signed you can see that the largest file is `sof-cnl.ri` and it's 545KB big and it's 96% full.

The ADSP memory is many, many orders of magnitude smaller compared to the systems this script runs on.
```
sof/tools/sof_ri_info/sof_ri_info.py v2.2.x/sof-v2.2-rc1/sof-cnl.ri 

  l2 hpsram                           0xbe000000 (3031040 + 114688  96.35% used)
```

enjiamai replied:
Please kindly correct me if I misunderstand your suggestion.  I tried this but it seems like it doesn't work well with large BUF_SIZE,.
```
    def receive_fw(self):
        header_sz = 80
        LARGE_BUF_SIZE = 16 * 1024 * 1024

        log.info("Receiving...")
        # The first contain the header and the front
        # data of the firmware
        d = self.request.recv(BUF_SIZE)

        # Unpacked the header data
        # Include size(4), filename(42), MD5(32) and fnamelen(2)
        header = d[:header_sz]
        total = d[header_sz:]
        s = struct.Struct(PACKET_HEADER_FORMAT_FW)
        fsize, fname, md5_tx_b, fnamelen = s.unpack(header)
        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, fnamelen:{fnamelen})')

        #receive all once in a large buffer.
        total += self.request.recv(LARGE_BUF_SIZE)  

        cnt = len(total)
        log.info(f"Done Receiving {cnt}.")
```
And it got 1368 most in our network environment.
```
INFO:cavs-fw:Req server start...
INFO:cavs-fw:Log server start...
INFO:cavs-fw:192.168.0.200 wrote: b'download'
INFO:cavs-fw:Receiving...
INFO:cavs-fw:size:114704, filename:b'zephyr.ri.9c23b0cc8538e7a58b9246a0b77e9b59', MD5:b'bbfe7524c87071c99fa737b4946a0ef9', fnamelen:42)
INFO:cavs-fw:Done Receiving 1368.
ERROR:cavs-fw:MD5 mismatch: bbfe7524c87071c99fa737b4946a0ef9 vs. 3d6f257012bf0f2a93428208b5d75a1a
ERROR:cavs-fw:Receive firmware failed.

```
So I suggest to keep the current way with enlarging the buffer size.

marc-hb replied:
Apparently we can't have nice things... there's apparently nothing like `recvall()` in the Python socket API, so you're right: the `while recv()` dance is required. Sorry for misleading you here.

enjiamai replied:
not at all, thank you so much for this suggestion!

marc-hb reviewed (changes requested):
Please prove me wrong but I think I found at least one error handling bug.

Also found a number of simplifications.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -41,27 +43,57 @@ def send_cmd(self, cmd):
+        values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'), flen)
+        log.info(f'filename:{fname}, size:{fsize}, md5:{md5_tx}, len {flen}')
+
+        s = struct.Struct('I 42s 32s H')
```
""convention" sounds like it's defined somewhere else but `'I 42s 32s H'` is defined only in this repo (which is great, it means it can be changed any time without breaking backward compatibility).

Please name this "magic value" / protocol header a name at the top of the file:

```
PACKET_HEADER_FORMAT_SOMETHING = 'I 42s 32s H'
```

enjiamai replied:
Good suggestion got it fixed. Thanks!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,65 @@ async def _main(server):
+        cnt = len(total)
+        s = struct.Struct('I 42s 32s H')
+        fsize, fname, md5_tx_b, pad = s.unpack(header)
+        log.info(f'size:{fsize}, filename:{fname}, MD5:{md5_tx_b}, pad:{pad})')
```
On the sending side this was `values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'), flen)` where `flen` was the filename length (before encoding). Why is becoming `pad`?

enjiamai replied:
please refer to above comment, thanks!

marc-hb reviewed (changes requested):
Just found another issue:
```
flen = len(fname)
...
values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'), flen)
```
This is sending the length of the _decoded_ filename. It will match the encoded filename only when the filename is ASCII. This is not a real problem yet because `flen` becomes `pad` on the receiving side (?) and is apparently not used for now. 


=> Too many problems designing your own network protocol, please switch to something standard like JSONRPC or Python Remote Objects ASAP - anything off the shelf really.

From https://docs.python.org/3/howto/sockets.html
> Sockets are used nearly everywhere, but are one of the most severely misunderstood technologies around. 

PS: not relevant here but just "for fun" https://unicode.org/reports/tr36/#Buffer_Overflows

marc-hb commented:
> => Too many problems designing your own network protocol, please switch to something standard like JSONRPC or Python Remote Objects ASAP - anything off the shelf really.


The example https://github.com/irmen/Pyro5/tree/master/examples/hugetransfer shows how to transfer a file over the network with practically no code at all. There's very little code there and most of it is for testing purposes.


> cavstool_client.py -s {hostname}:{req_port}:{log_port}

Please use a new `--argument` for this. Don't extend standard conventions; they're painful enough to parse already. The number of colons `:` is used as a trick to identify bracket-less IPv6 addresses which means `{hostname}:{req_port}:{log_port}` could be wrongly identified as an IPv6 address. 

https://bugzilla.mindrot.org/show_bug.cgi?id=1602
https://chromium-review.googlesource.com/c/chromiumos/third_party/autotest/+/312719/3/server/base_utils.py

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,65 @@ async def _main(server):
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while cnt < fsize:
```
Ok, got it, thanks. Fixed.

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -41,27 +43,57 @@ def send_cmd(self, cmd):
         # Send the FW to server
+        fname = os.path.basename(filename)
+        fsize = os.path.getsize(filename)
+        flen = len(fname)
```
You are right, the filename length is not used here because we have a fixed filename. At first I just want to make 4 bytes align. Let's remove them and keep it clean. Thanks.

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +660,65 @@ async def _main(server):
+        # receive the specified amount of bytes.
+        while cnt < fsize:
+            data = self.request.recv(BUF_SIZE)
+            if not data:
```
Please let me explain it: In the server-side, the recv() will be blocked until the client's socket.shutdown/close called. We hope the client can keep receiving the download result in the current connection (after completing MD5 check) from the server and return it to the west runner. So we only receive the specified bytes then keep running and inform the client.

For using python Remote Objects or JSONRPC and I think it's a good suggestion but I might need a little more time to realize it. Thank you.

marc-hb replied:
`not data` means the socket has been closed. Why would the socket be closed when there is still data to read? Only when something went wrong => exception.

Please describe what happens when you remove this `break`.

enjiamai replied:
hi @marc-hb , it seem it still work well if I remove the break.

enjiamai replied:
I think the break here can be remove.

marc-hb replied:
When I asked to remove the `break` it was only temporarily for testing purposes, not for the final code. You still need to check for disconnections otherwise you can have an infinite loop.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +662,62 @@ async def _main(server):
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while len(total) < fsize:
+            data = self.request.recv(LARGE_BUF_SIZE)
```
```suggestion
            data = self.request.recv(min(LARGE_BUF_SIZE, fsize - len(total))
```
otherwise we're back to the bug https://github.com/zephyrproject-rtos/zephyr/issues/46864 that this is supposed to fix.

From https://docs.python.org/3/howto/sockets.html
> One complication to be aware of: if your conversational protocol allows multiple messages to be sent back to back (without some kind of reply), and you pass recv an arbitrary chunk size, you may end up reading the start of a following message

enjiamai replied:
Good suggestion, applied it.

marc-hb reviewed (dismissed):
I think there are still some `recv()` issues.

PS: have you considered moving the port mapping commit to a different PR?

enjiamai commented:
Hi @marc-hb , I appreciate you review and give me many great suggestions and detailed reference, learn so much! And I am trying to address all of them as possible but one might not. I think for changing to using the JSONRPC or Python could be next step after this PR merged.

And according to your suggestion, I split the support for specifying port to another PR. https://github.com/zephyrproject-rtos/zephyr/pull/46972.  Because in one of our environment that is necessary. I put it there in another PR until I completed the --argument, such as --port-req and --port-log . ( At first I want change it this way, but it also need to change the west runner a little bit, and the documents also need to updated. So I choose the simplest way and try to make the arguments less. )

Thanks for spending much time to review this :)

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +662,60 @@ async def _main(server):
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+        LARGE_BUF_SIZE = 16 * 1024 * 1024
```
Looks like this is leftover from my bad `recvall()` suggestion (there is no `recvall` , sorry for misleading you). I think you can simplify and use 4K everywhere.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -25,8 +26,9 @@
 CMD_LOG_START = "start_log"
-CMD_LOG_STOP = "stop_log"
 CMD_DOWNLOAD = "download"
+PACKET_HEADER_FORMAT_FW = 'I 42s 32s H'
```
Add a comment with field names (and get rid of the unused `pad`?)

enjiamai replied:
Hi @marc-hb , how do you think I make it as "reserve"? Just in case there is new requirement in the future. thanks!

marc-hb replied:
You don't need to reserve anything: the only two scripts that speak this protocol both live in the same directory. If you need to change the format then you just go and do it; no need for backward compatibility. Using one script from one git version and the other script from a different git version is not supported: why would it?

Same thing when you drop all this low-level socket code and switch to something higher level :-)

enjiamai replied:
Fixed. Remove it, thanks!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +662,59 @@ async def _main(server):
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while len(total) < fsize:
+            data = self.request.recv(min(BUF_SIZE, fsize - len(total)))
```
When I asked to remove the `if not data; break` it was only temporarily for testing purposes. You still need to check for disconnections otherwise you can trigger an infinite loop. You need this:

```diff
--- a/soc/xtensa/intel_adsp/tools/cavstool.py
+++ b/soc/xtensa/intel_adsp/tools/cavstool.py
@@ -710,6 +710,8 @@ class adsp_request_handler(socketserver.BaseRequestHandler):
         # receive the specified amount of bytes.
         while len(total) < fsize:
             data = self.request.recv(min(BUF_SIZE, fsize - len(total)))
+            if not data:
+                raise EOFError("truncated firmware file")
             total += data
 

```

Here's how to test or reproduce the infinite loop:

```diff
--- a/soc/xtensa/intel_adsp/tools/cavstool_client.py
+++ b/soc/xtensa/intel_adsp/tools/cavstool_client.py
@@ -73,7 +73,15 @@ class cavstool_client():
             log.info(f'Sending...')
 
             total = self.sock.send(header_data)
-            total += self.sock.sendfile(f)
+
+            fw = f.read()
+            total += len(fw)
+
+            if False:
+                self.sock.send(fw) # Works fine
+            else:  # truncationg hangs forever; then infinite loop on the test device after
+                # interrupting cavstool_client.py with Ctrl-C
+                self.sock.send(fw[:-5])
 
             log.info(f"Done Sending ({total}).")

```

enjiamai replied:
That's fantastic! Nice catch! Thank you!

marc-hb reviewed (changes requested):
Thanks @enjiamai for the lighting fast updates despite the timezone differences.

I (finally) tested the current revision 5e95e5824f30b70d1 and it is indeed much more reliable! I will keep using it locally.

One last infinite loop problem on disconnections (see below) and then it should LGTM

PS: thanks for considering to drop bare sockets programming in the future; it is really too level and too time consuming. Let's keep that for Zephyr itself, no need to go that low level in Python test code :-)

marc-hb commented:
> Thanks for spending much time to review this :)

Don't thank me: I need this for various reasons and I spend a number of hours trying to understand why it didn't work. I'm very selfish ;-)

> Firmware in production will never use this and all final testing has to happen in a SOF binary with the kernel loaders.

Famous last words:
- https://github.com/thesofproject/sof-test/pull/897

> It might be time to split the code up so that a minimal "just run this file please" tool can continue to be maintained.

This will require copying more than one script to the DUT.

This will require SOF updates otherwise it will break https://github.com/thesofproject/sof/commit/c9ee1092965264 and SOF CI like this recently did:
- #46308

We should first "upgrade" SOF to deploy an entire subdirectory.

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +662,59 @@ async def _main(server):
+            data = self.request.recv(min(BUF_SIZE, fsize - len(total)))
+            total += data
+
+        cnt = len(total)
```
This is for following log output, or do you think we can also simplify it? 
```
        cnt = len(total)
        log.info(f"Done Receiving {cnt}.")
```

marc-hb replied:
Sorry I missed the usage in logging. Yes I would probably simplify it to something like `log.info(f"Done Receiving {len(total)}.")`. No big deal.

enjiamai replied:
Okok, update it to be more simplifed. thanks!

enjiamai commented:
Ok, @marc-hb , I so appreciate your review, even writing tests for it. :)

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -659,30 +662,60 @@ async def _main(server):
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 78
```
Move this next to `PACKET_HEADER_FORMAT_FW`

enjiamai replied:
Ok, fixed.

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -41,27 +44,47 @@ def send_cmd(self, cmd):
+
+            log.info(f"Done Sending ({total}).")
+
+            rck = self.sock.recv(BUF_SIZE).decode("utf-8")
```
Again you should not read `BUF_SIZE` but OK, here it's less likely to fail.

enjiamai replied:
ok, fixed.

marc-hb reviewed (dismissed):
Tested again and works fine. Good enough for a quick fix.

The commit message is not really accurate:

> soc: xtensa: refine the client-server CAVS tool

> 1. Improve the firmware transfer reliability. The sendfile API is
not stable and sometimes we got incorrect contents.

It should say something like:

```
soc: xtensa: partial FIX (not just "refine") of socket misuse 

Fix the most frequent occurence of the common `recv()` bug described here:

https://docs.python.org/3/howto/sockets.html#using-a-socket

> One complication to be aware of: if your conversational protocol allows multiple messages to be sent back to back (without some kind of reply), and you pass recv an arbitrary chunk size, you may end up reading the start of a following message. 

Partial fix for 46864

The longer term fix is to switch to a higher level API like Python Remote Objects


```

marc-hb commented:
> > It might be time to split the code up so that a minimal "just run this file please" tool can continue to be maintained.

> This will require (SOF CI) copying more than one script to the DUT.

Submitted in
- https://github.com/thesofproject/sof/pull/5968

enjiamai commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -2,20 +2,23 @@
 CMD_LOG_START = "start_log"
-CMD_LOG_STOP = "stop_log"
 CMD_DOWNLOAD = "download"
+PACKET_HEADER_FORMAT_FW = 'I 42s 32s'
```
ok, fixed.

enjiamai commented:
Hi @marc-hb , addressed your comments. Could you please re-review it, thanks!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool_client.py:
```diff
@@ -36,32 +45,52 @@ def send_cmd(self, cmd):
             self.sock.sendall(cmd.encode("utf-8"))
             log.info(f"Sent:     {cmd}")
-            ack = str(self.sock.recv(BUF_SIZE), "utf-8")
+            ack = str(self.sock.recv(CMD_SZ), "utf-8")
```
This does not make a lot of difference because commands have a variable and unknown size. So you could still be reading too much; still more or less the same bug. In this particular case I think it won't happen in practice because I believe you have only one RPC in flight at a time and all commands fit in a single Ethernet packet but it's still brittle.

To be fixed when when this gets "upgraded" away from bare sockets, reliability is massively better with this PR already.

enjiamai replied:
Agree with you, let's put it in the next update. Thank you!

marc-hb commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -24,9 +25,17 @@
 CMD_LOG_START = "start_log"
-CMD_LOG_STOP = "stop_log"
 CMD_DOWNLOAD = "download"
+CMD_SZ = 16
```
```suggestion
MAX_CMD_SZ = 16
```
Otherwise it looks like commands have constant size (which would be more correct, see below).

But don't do it now to save time. Change it only if another reviewer finds a serious issue and requests other changes.

enjiamai replied:
I got it, I was thinking to make the command length to be 16(with '\0') , but maybe we should change it in a higher level, that's way I did not make it fit totally.

enjiamai replied:
Got it fixed, thanks!

enjiamai commented:
Thank you for the review again! @marc-hb. :)

laurenmurphyx64 commented:
Looks good for WHL and APL in our HW pool, but both of our TGL are failing every time for some reason? Not sure what's going on there. Sorry to bring bad news so close to the end of your marathon... :disappointed: I haven't been able to power cycle either board yet, so maybe that's related?  

Well, as long as your other PR (https://github.com/zephyrproject-rtos/zephyr/pull/46889) supporting the old flasher gets merged (all you need to do is update the docs + make a note in cavstwist.h, I think), users of the HW pool can use that. So as long as this PR is working for SOF, maybe it's OK to merge this partial sockets fix PR and then start on the Python remote objects / similar fix...?

```
zephyr@localhost:~$ sudo ./cavs_server.py -s 192.168.0.142
INFO:cavs-fw:Req server start...
INFO:cavs-fw:Log server start...
INFO:cavs-fw:192.168.0.200 wrote: b'download'
INFO:cavs-fw:Receiving...
INFO:cavs-fw:size:65552, filename:b'zephyr.ri.62d7aca976509a94762dda506b0f8a82', MD5:b'4f41d117d93b0d8b6b155e37c60d7ff2'
INFO:cavs-fw:Done Receiving 65552.
INFO:cavs-fw:Firmware well received. Ready to download.
INFO:cavs-fw:192.168.0.200 wrote: b'start_log'
INFO:cavs-fw:Waiting for instruction...
INFO:cavs-fw:Loaded FW b'zephyr.ri.62d7aca976509a94762dda506b0f8a82' and running...
INFO:cavs-fw:Mapped PCI bar 0 of length 16384 bytes.
INFO:cavs-fw:Selected output stream 7 (GCAP = 0x1009701)
INFO:cavs-fw:Mapped PCI bar 4 of length 1048576 bytes.
INFO:cavs-fw:Detected cAVS 1.8+ hardware
INFO:cavs-fw:Trimming 16 bytes of extended manifest
INFO:cavs-fw:Resetting HDA device
INFO:cavs-fw:Stalling and Resetting DSP cores, ADSPCS = 0xf010000
INFO:cavs-fw:Powering down DSP cores, ADSPCS = 0xf010101
INFO:cavs-fw:Configuring HDA stream 7 to transfer firmware image
INFO:cavs-fw:type of mem is <class 'mmap.mmap'>
INFO:cavs-fw:Mapped 2M huge page at 0x23d000000 to contain 65536 bytes of firmware
INFO:cavs-fw:Filled the buffer descriptor list (BDL) for DMA.
INFO:cavs-fw:Starting DSP, ADSPCS = 0x101
INFO:cavs-fw:Unresetting DSP cores, ADSPCS = 0x1010000
INFO:cavs-fw:Running DSP cores, ADSPCS = 0x1010000
INFO:cavs-fw:Wait for ROM startup, ADSPCS = 0x1010000
INFO:cavs-fw:Sending IPC command, HIPIDR = 0x81004000
INFO:cavs-fw:Starting DMA, FW_STATUS = 0x5000001
INFO:cavs-fw:Waiting for firmware handoff, FW_STATUS = 0x5000001
WARNING:cavs-fw:Load failed?  FW_STATUS = 0x80000012
INFO:cavs-fw:cAVS firmware load complete
```

enjiamai commented:
> Looks good for WHL and APL in our HW pool, but both of our TGL are failing every time for some reason? Not sure what's going on there. Sorry to bring bad news so close to the end of your marathon... 😞 I haven't been able to power cycle either board yet, so maybe that's related?

Thanks you,  @laurenmurphyx64 , let me take a look right now.

enjiamai commented:
Hi @laurenmurphyx64 , I cannot get a TGL board in remote env right away, But from the log you show up, it seems like that is not a matching version of it. (The latest version has a MD5sum check.)  So I think I have to add a version info for it.

enjiamai commented:
Sorry, @laurenmurphyx64  I omitted some msg, it has MD5 checksum. My bad.

marc-hb commented:
> Looks good for WHL and APL in our HW pool, but both of our TGL are failing every time for some reason? 

What's in the logs; is the failure related to this in any way?

laurenmurphyx64 commented:
@marc-hb  Enjia tried it again on our TGL and didn't have any failures, even with a Twister run, so I think this might've just been user error on my part (there could've been processes running on the ports). I can't verify myself right now since I'm having some non-CAVS related issues with our HW setup, but it sounds like this was a false alarm.

enjiamai commented:
> @marc-hb Enjia tried it again on our TGL and didn't have any failures, even with a Twister run, so I think this might've just been user error on my part (there could've been processes running on the ports). I can't verify myself right now since I'm having some non-CAVS related issues with our HW setup, but it sounds like this was a false alarm.
``` 
WARNING:cavs-fw:Load failed?  FW_STATUS = 0x80000012
INFO:cavs-fw:cAVS firmware load complete
``` 
Hi @laurenmurphyx64 , I can reproduce it only when the signing key is incorrect for TGL: (because the TGL use  otc_xxx_3k.key but WHL and APL use otc_xxx.key without 3k-suffix.) could you please double check your hardware map file it when you got a chance, thanks!

I believe this PR can make the CAVS client/server work quite stable. And apply this PR with https://github.com/zephyrproject-rtos/zephyr/pull/46972, it can also work quite stable in the port mapping environment. :)
